core/server/models/post.js

```diff
@@ -676,24 +678,56 @@ Post = ghostBookshelf.Model.extend({
             return unsafeAttrs[attr] && unsafeAttrs[attr] !== postModel.get(attr);
         }
 
-        function actorIsAuthor(loadedPermissions) {
-            return loadedPermissions.user && _.some(loadedPermissions.user.roles, {name: 'Author'});
-        }
-
         function isOwner() {
             return unsafeAttrs.author_id && unsafeAttrs.author_id === context.user;
         }
 
-        if (actorIsAuthor(loadedPermissions) && action === 'edit' && isChanging('author_id')) {
-            hasUserPermission = false;
-        } else if (actorIsAuthor(loadedPermissions) && action === 'add') {
+        function isCurrentOwner() {
+            return context.user === postModel.get('author_id');
+        }
+
+        function isPublished() {
+            return unsafeAttrs.status && unsafeAttrs.status !== 'draft';
+        }
+
+        function isDraft() {
+            return postModel.get('status') === 'draft';
+        }
+
+        isContributor = baseUtils.actorIs(loadedPermissions.user, 'Contributor');
+        isAuthor = baseUtils.actorIs(loadedPermissions.user, 'Author');
+        isEdit = (action === 'edit');
+        isAdd = (action === 'add');
+        isDestroy = (action === 'destroy');
+
+        if (isContributor && isEdit) {
+            // Only allow contributor edit if neither status or author id are changing, and the post is a draft post
+            hasUserPermission = !isChanging('status') && !isChanging('author_id') && isDraft();
+        } else if (isContributor && isAdd) {
+            // If adding, make sure it's a draft post and has the correct ownership
+            hasUserPermission = !isPublished() && isOwner();
+        } else if (isContributor && isDestroy) {
+            // If destroying, only allow contributor to destroy their own draft posts
+            hasUserPermission = isCurrentOwner() && isDraft();
+        } else if (isAuthor && isEdit) {
+            // Don't allow author to change author ids
+            hasUserPermission = isCurrentOwner() && !isChanging('author_id');
+        } else if (isAuthor && isAdd) {
+            // Make sure new post is authored by the current user
             hasUserPermission = isOwner();
         } else if (postModel) {
-            hasUserPermission = hasUserPermission || context.user === postModel.get('author_id');
+            hasUserPermission = hasUserPermission || isCurrentOwner();
+        }
+
+        if (isContributor) {
+            // Note: at the moment primary_tag is a computed field,
+            // meaning we don't add it to this list. However, if the primary_tag
+            // ever becomes a db field rather than a computed field, add it to this list
+            result.excludedAttrs = ['tags'];
```

Comment:
Okay i think i see why you have added this approach, because `tags` is a nested relation and it's **not available**, because we fetch the model without relations. 

> this.findOne({id: postModelOrId, status: 'all'})

Excluding tags from permission checks makes no sense to me 🤔 The client sends the payload including tags. If you are a contributor role and you change them, you should get a permission error? Otherwise we have two different concepts here:

1. return permission error if you change the author id
2. do not return a permission error if you send tags

Maybe share a little bit of your thinking with me 👍

Reply:
This was part of a discussion @ErisDS and I had about the original spec. The eventual goal is for contributors to be able to add _existing_ tags to a post, however the work required to get that permissions checking into the model layer would make this feature take a lot longer.

It should be possible (I think) to throw a permissions error when tags are passed and the user is a contributor, but the original spec said we'd just ignore any passed tags.

Reply:
Hmm but as long as we don't add the support to add existing tags for contributors, i would prefer sticking to one concept?

Reply:
> but the original spec said we'd just ignore any passed tags.

Can't find that in the contributor spec 🤔 . It just says:

> Preventing contributors from being able to add any tags to a post is a shortcut that makes it simple to implement this spec.

Reply:
> . If any tags exist on the post (e.g. added by an editor or something pre-publish) and then the contributor tries to save the post, an error will be thrown because there are tags in the request data

What exactly did not work? If the tags did not change, it should pass?

Reply:
That's where the tricky part is I think - we have to fetch the tags relation _just_ to see if the contributor can edit/add. It makes the logic of this function a _whole_ lot more tricky I think too, because then we either have to do a deep compare (which may or may not work depending on what fields Ghost Admin includes as opposed to what the server has), or compare id's/etc, which means adding more code & tests.

The reason it's not in the spec is @ErisDS and I had a chat prior to me starting work on this about the tags stuff specifically, and decided that just excluding the tags from the result was the easiest option for the first pass on the contributor role.

Reply:
IMO the path of least resistance is correct here - whatever that is 😄

Reply:
Okay then let's keep as is. 

@acburdine 
Could you pls extend your comment to say that if we are adding the feature to allow adding *existing* tags, the concept of `excludingAttrs` has to be removed? (except of we are re-using the new concept for other cases)